3.analyze_data/utils/isolation_forest_data_feature_importance.py

```diff
@@ -0,0 +1,148 @@
+from collections import defaultdict
+from typing import Optional, Sequence, Union
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+from sklearn.tree import DecisionTreeRegressor, _tree
+
+
+class IsoforestFeatureImportance:
+    # Computes Isolation Forest (Scikit-learn) feature importances from a subset of data.
+
+    def __init__(
+        self,
+        _estimators: Sequence[DecisionTreeRegressor],
+        _morphology_data: pd.DataFrame,
+    ):
+        """
+        Parameters
+        ----------
+        _estimators: Scikit-learn decision tree regressor estimators.
+        _morphology_data: Data with only numerical morphology data (without metadata).
+        """
+
+        self._estimators = _estimators
+        self._morphology_data = _morphology_data
+        self._isoforest_importances = None
+
+    @property
+    def isoforest_importances(self):
+        if self._isoforest_importances is None:
+            raise ValueError("isoforest_importances have not been computed")
+
+        return self._isoforest_importances
+
+    def save_tree_feature_importances(
+        self,
+        _tree_obj: _tree.Tree,
+        _leaf_id: int,
+        _sample_idx: int,
+    ) -> dict[dict[str, float]]:
+        """
+        Aggregates and then returns feature importances per tree.
+
+        Parameters
+        _tree_obj: Object for traversing tree.
+        _leaf_id: Leaf node for the sample in the tree.
+        _sample_idx: Sample location in the dataframe.
+        """
+
+        node_id = 0  # Start at the root node
+        depth = 0
+        feature_importances = defaultdict(list)
+        morphology_features = self._morphology_data.iloc[_sample_idx].copy()
+
+        while node_id != _leaf_id:
+            feature_idx = _tree_obj.feature[node_id]
+
+            if feature_idx >= 0:  # Ignore leaf nodes (-2)
+                """
+                Samples are often easier to seperate with less data.
+                Therefore, earlier features have greater importances.
+                """
```

Comment:
it feels a bit strange to see a docstring here - why not add this as a more traditional comment or move to the function docstring?